pkg/kubecost/summaryallocation.go

```diff
@@ -0,0 +1,1155 @@
+	// used to compute idle coefficients, based on the ratio of any given
+	// allocation's per-resource cost to the allocated per-resource totals.
+	var allocTotals map[string]*ResourceTotals
+	if options.IdleByNode {
+		allocTotals = options.AllocationResourceTotalsStore.GetResourceTotalsByNode(*sas.Window.Start(), *sas.Window.End())
+		if allocTotals == nil {
+			// TODO
+			log.Warningf("SummaryAllocation: nil allocTotals by node for %s", sas.Window)
+		}
+	} else {
+		allocTotals = options.AllocationResourceTotalsStore.GetResourceTotalsByCluster(*sas.Window.Start(), *sas.Window.End())
+		if allocTotals == nil {
+			// TODO
+			log.Warningf("SummaryAllocation: nil allocTotals by cluster for %s", sas.Window)
+		}
+	}
+
+	// Instantiate filtered totals map if filters have been applied and there
+	// are un-shared idle allocations in the result set. These will be
+	// populated in step (#) and used to "filter" out the correct proportion
+	// of idle costs, given that non-idle allocations have been filtered.
+	var filteredTotals map[string]*ResourceTotals
+	if len(aggSet.idleKeys) > 0 && len(options.FilterFuncs) > 0 {
+		filteredTotals = make(map[string]*ResourceTotals, len(aggSet.idleKeys))
+
+		// If costs are shared, record those resource totals here so that idle
+		// for the shared resources gets included.
+		for key, rt := range sharedResourceTotals {
+			if _, ok := filteredTotals[key]; !ok {
+				filteredTotals[key] = &ResourceTotals{}
+			}
+
+			// TODO do we need PV, Network, LoadBalancer here?
+
+			filteredTotals[key].CPUCost += rt.CPUCost
+			filteredTotals[key].GPUCost += rt.GPUCost
+			filteredTotals[key].RAMCost += rt.RAMCost
+		}
+	}
+
+	sharingCoeffs := map[string]float64{}
+
+	// (#) Distribute idle cost, if sharing
+	// (#) Distribute tenancy costs, if sharing
+	// (#) Aggregate
+	// TODO better comment
+	for _, sa := range sas.SummaryAllocations {
+		// Generate key to use for aggregation-by-key and allocation name
+		key := sa.generateKey(aggregateBy, options.LabelConfig)
+
+		// TODO comment
+		// Do this before adding idle cost
+		if options.ShareSplit == ShareEven {
+			sharingCoeffs[key] += 1.0
+		} else {
+			sharingCoeffs[key] += sa.TotalCost()
+		}
+
+		// Distribute idle allocations according to the idle coefficients
+		// NOTE: if idle allocation is off (i.e. ShareIdle == ShareNone) then
+		// all idle allocations will be in the aggSet at this point, so idleSet
+		// will be empty and we won't enter this block.
+		if len(idleSet.SummaryAllocations) > 0 {
+			// Distribute idle allocations by coefficient per-idleId, per-allocation
+			for _, idle := range idleSet.SummaryAllocations {
+				var key string
+
+				// Only share idle allocation with current allocation (sa) if
+				// the relevant property matches (i.e. Cluster or Node,
+				// depending on which idle sharing option is selected)
+				if options.IdleByNode {
+					if idle.Properties.Node != sa.Properties.Node {
+						continue
```

Comment:
Any danger here if there are two clusters with identically-named nodes?

Reply:
Yes, and I believe there is a TODO comment later on somewhere about exactly that. Good question.